### PR TITLE
Allow constructor Verification

### PR DESCRIPTION
**Pull request by Roaders**

### CODE REVIEW

**Comment by Davidhanson90** [comment]
Can we add some docs for this new feature?

File: main/mock/contracts.ts [hunk 1]
```diff
@@ -38,34 +38,55 @@ export type FunctionCallLookup<T, C extends ConstructorFunction<T>, U extends Lo
...
     : U extends SetterTypes
-    ? [VerifierTarget<T, C, U>[K]]
+    ? [LookupFunction<T, C, U, K>]
     : [];
+
 export type FunctionParams<T> = T extends (...args: infer P) => any ? P : never;
 
 export type ConstructorFunction<T> = (abstract new (...args: any[]) => T) | (new (...args: any[]) => T);
 
+export type ConstructorParams<T extends ConstructorFunction<any>> = T extends abstract new (
```
Comment (Davidhanson90) [inline main/mock/contracts.ts:65]: Need to start documenting the types.  RAbstract etc

File: readme.md [hunk 2]
```diff
@@ -2,7 +2,6 @@
 ![Lifecycle Active](https://badgen.net/badge/Lifecycle/Active/green)
 ![npm](https://img.shields.io/npm/v/@morgan-stanley/ts-mocking-bird)
 [![Build Status](https://github.com/morganstanley/ts-mocking-bird/actions/workflows/build.yml/badge.svg)](https://github.com/Roaders/ts-mocking-bird/actions/workflows/build.yml)
-[![codecov](https://codecov.io/gh/morganstanley/ts-mocking-bird/branch/master/graph/badge.svg)](https://codecov.io/gh/morganstanley/ts-mocking-bird)
```
Comment (Davidhanson90) [inline readme.md:5]: need docs

File: main/mock/verifiers.ts [hunk 3]
```diff
@@ -143,9 +167,21 @@ export function verifyFunctionCalled<T, C extends ConstructorFunction<T>, U exte
...
     let errorMessageDescription: string;
 
-    const functionCalls: LookupParams<T, C, U, any>[] | undefined = getLookup(mock, type)[functionName];
+    const functionCallsLookup = getLookup(mock, type);
+
+    console.log(`functionCallsLookup`, { functionCallsLookup });
+
+    const functionCalls: LookupParams<T, C, U, any>[] | undefined = functionCallsLookup[functionName];
+
+    console.log(`functionCalls`, { functionCalls });
```
Comment (Davidhanson90) [inline main/mock/verifiers.ts]: Remove th console logs.

File: readme.md [hunk 4]
```diff
@@ -117,6 +116,19 @@ expect(
...
 ```
 
+### Verify Constructor Parameters
+
+```ts
+const mockInstance = Mock.create<ClassWithInstanceArgument, typeof ClassWithInstanceArgument>().setup(
+            setupConstructor(),
+        );
+
+new mockInstance.mockConstructor(serviceInstance);
```
Comment (Davidhanson90) [inline readme.md:126]: mockConstructor?  How does this fit with other semantics?  For example mock.setupFunction.withParameters()

I think the ergonmics of the API are often vague... not sure what approach should be taken for verification.

Comment (Roaders) [reply readme.md:126]: it has always been there. `mockedThing.mock` is the mocked instance `mockedThing.mockConstructor` is the mock constructor.
The equivalent of `mock.withFunction().withParameters()` is `mock.withConstructor().withParameters()` and `mock.setupFunction()` is `mock.setupConstrucotr()`

Comment (Davidhanson90) [reply readme.md:126]: so there ia mock.mockFunction('functionName') as well? If I am performing mock verification I would expect to see verifcation functions to help with verifying constructors, functions and properties. I would expect them to use the same naming. I think the problem arises when you have the ability to use the mock instance like this

mock.setupFunction('funtionName)

Should the setups only be available via the mock.setup method and the instance methods are for verifications.

mock.mockFunction('functionNamer')

**Review by Davidhanson90 — APPROVED** [review]

**Pull request merged**